gitlab/client_repositories_org.go

```diff
@@ -0,0 +1,162 @@
+	}
+	return newGroupProject(c.clientContext, apiObj, ref), nil
+}
+
+// List all repositories in the given organization.
+//
+// List returns all available repositories, using multiple paginated requests if needed.
+func (c *OrgRepositoriesClient) List(ctx context.Context, ref gitprovider.OrganizationRef) ([]gitprovider.OrgRepository, error) {
+	// Make sure the OrganizationRef is valid
+	if err := validateOrganizationRef(ref, c.domain); err != nil {
+		return nil, err
+	}
+
+	// GET /orgs/{org}/repos
+	apiObjs, err := c.c.ListGroupProjects(ctx, ref.Organization)
+	if err != nil {
+		return nil, err
+	}
+
+	// Traverse the list, and return a list of OrgRepository objects
+	repos := make([]gitprovider.OrgRepository, 0, len(apiObjs))
+	for _, apiObj := range apiObjs {
+		// apiObj is already validated at ListOrgRepos
+		repos = append(repos, newGroupProject(c.clientContext, apiObj, gitprovider.OrgRepositoryRef{
+			OrganizationRef: ref,
+			RepositoryName:  apiObj.Name,
+		}))
+	}
+	return repos, nil
+}
+
+// Create creates a repository for the given organization, with the data and options.
+//
+// ErrAlreadyExists will be returned if the resource already exists.
+func (c *OrgRepositoriesClient) Create(ctx context.Context, ref gitprovider.OrgRepositoryRef, req gitprovider.RepositoryInfo, opts ...gitprovider.RepositoryCreateOption) (gitprovider.OrgRepository, error) {
+	// Make sure the RepositoryRef is valid
+	if err := validateOrgRepositoryRef(ref, c.domain); err != nil {
+		return nil, err
+	}
+
+	apiObj, err := createProject(ctx, c.c, ref, ref.Organization, req, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return newGroupProject(c.clientContext, apiObj, ref), nil
+}
+
+// Reconcile makes sure the given desired state (req) becomes the actual state in the backing Git provider.
+//
+// If req doesn't exist under the hood, it is created (actionTaken == true).
+// If req doesn't equal the actual state, the resource will be updated (actionTaken == true).
+// If req is already the actual state, this is a no-op (actionTaken == false).
+func (c *OrgRepositoriesClient) Reconcile(ctx context.Context, ref gitprovider.OrgRepositoryRef, req gitprovider.RepositoryInfo, opts ...gitprovider.RepositoryReconcileOption) (gitprovider.OrgRepository, bool, error) {
+	// First thing, validate and default the request to ensure a valid and fully-populated object
+	// (to minimize any possible diffs between desired and actual state)
+	if err := gitprovider.ValidateAndDefaultInfo(&req); err != nil {
+		return nil, false, err
+	}
+
+	actual, err := c.Get(ctx, ref)
+	if err != nil {
+		// Create if not found
+		if errors.Is(err, gitprovider.ErrNotFound) {
+			resp, err := c.Create(ctx, ref, req, toCreateOpts(opts...)...)
+			return resp, true, err
+		}
+
+		// Unexpected path, Get should succeed or return NotFound
+		return nil, false, err
+	}
+	// Run generic reconciliation
+	if *actual.Get().DefaultBranch == "" {
+		*actual.Get().DefaultBranch = "master"
```

Comment:
Does the gitlab server raise an error if default branch is blank?

Reply:
This won't return an error but it's easier to translate it to the default branch once a repo is created in the project.
Removing this if clause though as it is covered by the validation of the API response, thanks for pointing it out!